lib/src/set.dart

```diff
@@ -0,0 +1,341 @@
+import 'collections.dart';
+
+/// RealmSet is a collection that contains no duplicate elements.
+abstract class RealmSet<T extends Object?> extends SetBase<T> with RealmEntity implements Finalizable {
+  RealmObjectMetadata? _metadata;
+
+  /// Gets a value indicating whether this collection is still valid to use.
+  ///
+  /// Indicates whether the [Realm] instance hasn't been closed,
+  /// if it represents a to-many relationship
+  /// and it's parent object hasn't been deleted.
+  bool get isValid;
+
+  /// Creates an unmanaged RealmSet from [items]
+  factory RealmSet(Set<T> items) => UnmanagedRealmSet(items.toSet());
+
+  /// Allows listening for changes when the contents of this collection changes.
+  Stream<RealmSetChanges<T>> get changes;
+
+  /// Returns the element of type `T` at the specified [index].
+  ///
+  /// Note that elements in a RealmSet move around arbitrarily when other elements are
+  /// inserted/removed.
+  @override
+  T elementAt(int index) => super.elementAt(index);
+
+  /// Whether [value] is in the set.
+  /// ```dart
+  /// final characters = <String>{'A', 'B', 'C'};
+  /// final containsB = characters.contains('B'); // true
+  /// final containsD = characters.contains('D'); // false
+  /// ```
+  @override
+  bool contains(covariant T element); //Note: explicitly overriding contains() to change parameter type
+
+  /// If an object equal to [object] is in the set, return it.
+  ///
+  /// Checks whether [object] is in the set, like [contains], and if so,
+  /// returns the object in the set, otherwise returns `null`.
+  ///
+  /// If the equality relation used by the set is not identity,
+  /// then the returned object may not be *identical* to [object].
+  /// Some set implementations may not be able to implement this method.
+  /// If the [contains] method is computed,
+  /// rather than being based on an actual object instance,
+  /// then there may not be a specific object instance representing the
+  /// set element.
+  /// ```dart
+  /// final characters = <String>{'A', 'B', 'C'};
+  /// final containsB = characters.lookup('B');
+  /// print(containsB); // B
+  /// final containsD = characters.lookup('D');
+  /// print(containsD); // null
+  /// ```
+  @override
+  T? lookup(covariant T element); //Note: explicitly overriding lookup() to change parameter type
+
+  /// Removes [value] from the set.
+  ///
+  /// Returns `true` if [value] was in the set, and `false` if not.
+  /// The method has no effect if [value] was not in the set.
+  /// ```dart
+  /// final characters = <String>{'A', 'B', 'C'};
+  /// final didRemoveB = characters.remove('B'); // true
+  /// final didRemoveD = characters.remove('D'); // false
+  /// print(characters); // {A, C}
+  /// ```
+  @override
+  bool remove(covariant T value); //Note: explicitly overriding remove() to change parameter type
+
+  /// Clear the set and if it is a set of Realm objects, delete all realm objects from the realm.
+  void deleteAll();
```

Comment:
That's weird - does it really delete the objects from Realm? I don't believe such an API belongs on the set - we don't have it for list and I don't think we expose a similar API in any of the other SDKs.

Reply:
Well the test expectations show that objects are indeed delete from the realm and this is exposed in the C-API. 
I don't see why not adding it. Perhaps we can extends List to support it as well later. I think it's a good usability improvement. Don't know how much we should sync on this. Perhaps we could leave it and see if it is requested for the other SDKs.

Reply:
I don't think it's a good idea to have it - generally, we want to expose operations that mutate the Realm on the Realm itself. We already have `Realm.deleteMany` - if we want to optimize performance we could detect that the iterable is a RealmSet and call the C API to delete all elements. But having both `clear` and `deleteAll` on RealmSet is not good design.